src/librustc_typeck/check/closure.rs

```diff
@@ -32,6 +32,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         debug!("check_expr_closure(expr={:?},expected={:?})",
                expr,
                expected);
+        // FIXME: See if expected_kind here can impact generators
```

Comment:
We could change the code to infer some expected_sig for generators too. I changed the FIXME to state that.